demos/twa-webview-fallback/src/main/java/com/google/browser/examples/twawebviewfallback/WebViewFallbackActivity.java

```diff
@@ -0,0 +1,106 @@
+import android.webkit.WebSettings;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import android.widget.FrameLayout;
+import android.widget.Toast;
+
+public class WebViewFallbackActivity extends AppCompatActivity {
+    public static final String KEY_LAUNCH_URI =
+            "com.google.browser.examples.twawebviewfallback.WebViewFallbackActivity.KEY_LAUNCH_URL";
+
+    private Uri launchUrl;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        this.launchUrl = this.getIntent().getParcelableExtra(KEY_LAUNCH_URI);
+
+        FrameLayout webViewContainer = new FrameLayout(this);
+        WebView webView = new WebView(this);
+        webView.setWebViewClient(createWebViewClient());
+        webViewContainer.addView(webView,
+                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
+
+        WebSettings webSettings = webView.getSettings();
+        webSettings.setJavaScriptEnabled(true);
+
+        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(
+                FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
+
+        setContentView(webViewContainer,layoutParams);
+        webView.loadUrl(launchUrl.toString());
+    }
+
+    private WebViewClient createWebViewClient() {
+        return new WebViewClient() {
+            @Override
+            public boolean onRenderProcessGone(
+                    WebView view, RenderProcessGoneDetail detail) {
+                ViewGroup vg = (ViewGroup)view.getParent();
+
+                // Remove crashed WebView from the hierarchy
+                // and ensure it is destroyed.
+                vg.removeView(view);
+                view.destroy();
+
+                // Create a new instance, and ensure it also
+                // handles crashes - in this case, re-using
+                // the current WebViewClient
+                WebView webView = new WebView(view.getContext());
+                webView.setWebViewClient(this);
+                vg.addView(webView);
+
+                // With the crash recovered, decide what to do next.
+                // We are sending a toast and loading the origin
+                // URL, in this example.
+                Toast.makeText(view.getContext(), "Recovering from crash",
+                        Toast.LENGTH_LONG).show();
+                webView.loadUrl(launchUrl.toString());
+                return true;
+            }
+
+            @Override
+            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+                Uri launchUrl = WebViewFallbackActivity.this.launchUrl;
+                Uri navigationUrl = request.getUrl();
+
+                // If the user is navigation to a different origin, use CCT to handle the navigation
+                if (!launchUrl.getScheme().equals(navigationUrl.getScheme())
```

Comment:
Probably outside the scope of this initial CL, but I'd be very interested to see if the OAuth dance would work. WebView Fallback -> CCT -> Custom Scheme -> Launch TWA -> WebView Fallback.

Reply:
+1 - I don't think it's going to work right now. My current goal is to add a naive implementation that won't handle more complex dances (and then iterate)